### PR TITLE
core: expose service worker url on service worker audit

### PR DESCRIPTION
**Pull request by adrianaixba**
for #9683

exposing service worker url, and making it available on the service worker audit

### CODE REVIEW

File: lighthouse-core/audits/service-worker.js [hunk 1]
```diff
@@ -73,23 +73,31 @@ class ServiceWorker extends Audit {
...
-   * @return {string|undefined}
+   * @return {Array<string> | undefined}
    */
   static getControllingScopeUrl(matchingSWVersions, registrations, pageUrl) {
     // Find the normalized scope URLs of possibly-controlling SWs.
-    const matchingScopeUrls = matchingSWVersions
-      .map(v => registrations.find(r => r.registrationId === v.registrationId))
-      .filter(/** @return {r is LH.Crdp.ServiceWorker.ServiceWorkerRegistration} */ r => !!r)
-      .map(r => new URL(r.scopeURL).href);
+    const serviceWorkerUrls = new Map();
```
Comment (paulirish) [inline lighthouse-core/audits/service-worker.js:80]: we've been slowly adopting a _foos_**By**_Bars_ convention for our map names.. i think it'd help here... like `scriptUrlByScopeUrl`

File: lighthouse-core/audits/service-worker.js [hunk 2]
```diff
@@ -73,23 +73,31 @@ class ServiceWorker extends Audit {
...
+
+    // Populate serviceWorkerUrls map with the scriptURLs and scopeUrls of matchingSWVersions and registrations
+    matchingSWVersions.forEach(function(version) {
+      const tempRegistration = registrations.find(r => r.registrationId === version.registrationId);
```
Comment (paulirish) [inline lighthouse-core/audits/service-worker.js:84]: `matchedRegistration` ?  this signal is just that we have a match... and thus can continue, yeah?

File: lighthouse-core/audits/service-worker.js [hunk 3]
```diff
@@ -73,23 +73,31 @@ class ServiceWorker extends Audit {
    * @param {Array<LH.Crdp.ServiceWorker.ServiceWorkerVersion>} matchingSWVersions
    * @param {Array<LH.Crdp.ServiceWorker.ServiceWorkerRegistration>} registrations
    * @param {URL} pageUrl
-   * @return {string|undefined}
+   * @return {Array<string> | undefined}
    */
   static getControllingScopeUrl(matchingSWVersions, registrations, pageUrl) {
```
Comment (paulirish) [inline lighthouse-core/audits/service-worker.js:78]: now that this returns more than just the scope url we should rename it... perhaps `getControllingServiceWorker`

File: lighthouse-core/audits/service-worker.js [hunk 4]
```diff
@@ -73,23 +73,31 @@ class ServiceWorker extends Audit {
...
+      }
+    });
 
     // Find most-specific applicable scope, the one controlling the page.
     // See https://w3c.github.io/ServiceWorker/v1/#scope-match-algorithm
-    const pageControllingScope = matchingScopeUrls
-      .filter(scopeUrl => pageUrl.href.startsWith(scopeUrl))
-      .sort((scopeA, scopeB) => scopeA.length - scopeB.length)
+    const pageControllingUrls = [...serviceWorkerUrls]
+      /* converts map to array, to properly filer and sort according to scopeUrl */
```
Comment (paulirish) [inline lighthouse-core/audits/service-worker.js:95]: using a Map() is good, but right now it pretty quickly is thrown away and we turn it into the `[['scope1', 'scriptA'],['scope2','scriptB']]` representation. So we don't get much value out of it.

and separately the current return value is an array of strings, but you kinda have to know which is which.. I'd feel better making it explicit.


i remember in an earlier version you had named properties. i think within the `matchingScopeUrls` functional methods you ended up making a `Array<{scopeUrl: string, scriptUrl: string}>` or something

@adrianaixba  you were right and that was better. :) i misled you as i thought doing that wasn't necessary. mea cupla.

so yeah  looking at it now.. the best return value for this `getControllingServiceWorker()` method would probably be `{{scopeUrl: string, scriptUrl: string}}`, so we might as well build that from the start.

File: lighthouse-core/test/audits/service-worker-test.js [hunk 5]
```diff
@@ -193,7 +197,9 @@ describe('Offline: service worker audit', () => {
     const manifest = {start_url: finalUrl};
 
     const output = ServiceWorker.audit(createArtifacts(swOpts, finalUrl, manifest));
-    assert.deepStrictEqual(output, {score: 1});
+    assert.deepStrictEqual(output.score, 1);
+    assert.deepStrictEqual(output.details.items[0].scopeURL, 'https://example.com/project/');
+    assert.deepStrictEqual(output.details.items[0].scriptURL, 'https://example.com/project/sw.js');
```
Comment (paulirish) [inline lighthouse-core/test/audits/service-worker-test.js:202]: this is satisfying. and exactly what the test should have been asserting beforehand anyway. without this assertion (`/project/sw` is the controlling sw),  the test was kinda useless.

File: lighthouse-core/audits/service-worker.js [hunk 6]
```diff
@@ -73,23 +73,33 @@ class ServiceWorker extends Audit {
...
+   * @return {{scopeUrl: string; scriptUrl: string} | undefined}
    */
-  static getControllingScopeUrl(matchingSWVersions, registrations, pageUrl) {
+  static getControllingServiceWorker(matchingSWVersions, registrations, pageUrl) {
     // Find the normalized scope URLs of possibly-controlling SWs.
-    const matchingScopeUrls = matchingSWVersions
-      .map(v => registrations.find(r => r.registrationId === v.registrationId))
-      .filter(/** @return {r is LH.Crdp.ServiceWorker.ServiceWorkerRegistration} */ r => !!r)
-      .map(r => new URL(r.scopeURL).href);
+    /** @type {{ scopeUrl: string; scriptUrl: string; }[]} */
```
Comment (connorjclark) [inline lighthouse-core/audits/service-worker.js:80]: nit: prefer `Array<...>`, especially for complex types

there's a bit of an unofficial war between me, @patrickhulce and @brendankenny about when to use `Array<...>` vs `...[]` :) I prefer using `[]` for simple, non-compound types like `string[]` or `LH.Artifact.Blah[]`, but others prefer always using it. eventually you may pick a side ...

File: lighthouse-core/audits/service-worker.js [hunk 7]
```diff
@@ -73,23 +73,33 @@ class ServiceWorker extends Audit {
...
+    const scriptByScopeUrlList = [];
+
+    // Populate serviceWorkerUrls map with the scriptURLs and scopeUrls of matchingSWVersions and registrations
+    matchingSWVersions.forEach(function(version) {
```
Comment (connorjclark) [inline lighthouse-core/audits/service-worker.js:84]: nit: prefer `for (const blah of blahs)` over `blahs.forEach`.

File: lighthouse-core/audits/service-worker.js (hunk 6 above)
Comment (patrickhulce) [reply lighthouse-core/audits/service-worker.js:80]: for the record I also have absolutely no problem with `string[]` or `number[]` :) 

I think we all agree when the types grow to be more complex (I'd probably switch earlier than @connorjclark and use `Array<LH.Nested.Thing>`) the `Array<T>` form is easier to read 👍

File: lighthouse-core/audits/service-worker.js [hunk 8]
```diff
@@ -73,23 +73,33 @@ class ServiceWorker extends Audit {
...
+   * @return {{scopeUrl: string; scriptUrl: string} | undefined}
    */
-  static getControllingScopeUrl(matchingSWVersions, registrations, pageUrl) {
+  static getControllingServiceWorker(matchingSWVersions, registrations, pageUrl) {
     // Find the normalized scope URLs of possibly-controlling SWs.
-    const matchingScopeUrls = matchingSWVersions
-      .map(v => registrations.find(r => r.registrationId === v.registrationId))
-      .filter(/** @return {r is LH.Crdp.ServiceWorker.ServiceWorkerRegistration} */ r => !!r)
-      .map(r => new URL(r.scopeURL).href);
+    /** @type {Array<{ scopeUrl: string; scriptUrl: string;}>} */
```
Comment (connorjclark) [inline lighthouse-core/audits/service-worker.js:80]:
```suggestion
    /** @type {Array<{scopeUrl: string; scriptUrl: string}>} */
```

File: lighthouse-core/audits/service-worker.js [hunk 9]
```diff
@@ -73,23 +73,33 @@ class ServiceWorker extends Audit {
...
+      }
+    }
 
     // Find most-specific applicable scope, the one controlling the page.
     // See https://w3c.github.io/ServiceWorker/v1/#scope-match-algorithm
-    const pageControllingScope = matchingScopeUrls
-      .filter(scopeUrl => pageUrl.href.startsWith(scopeUrl))
-      .sort((scopeA, scopeB) => scopeA.length - scopeB.length)
+    const pageControllingUrls = scriptByScopeUrlList
+      .filter(urlPair => pageUrl.href.startsWith(urlPair.scopeUrl.toString()))
```
Comment (Beytoven) [inline lighthouse-core/audits/service-worker.js:98]: Nit: I believe `urlPair.scopeUrl` is already of type `string` so the `.toString()` isn't needed.

File: lighthouse-core/audits/service-worker.js [hunk 10]
```diff
@@ -73,23 +73,33 @@ class ServiceWorker extends Audit {
...
+      .sort((scopeA, scopeB) => scopeA.scopeUrl.length - scopeB.scopeUrl.length)
```
Comment (paulirish) [inline lighthouse-core/audits/service-worker.js:99]: extra nit:   we were calling this `urlPair` when it was an array of strings, but now its a typed object... so the identity and type is pretty clear. 

also we call this obj `urlPair` here on 98 but then `scopeA/B` on 99.  we should keep it consistent.

in these sorts of cases we often go for short/singleletter var names...  like this: 
https://github.com/GoogleChrome/lighthouse/blob/960d7c70a14111eefbe6cc78edb0a46b0512cbc7/lighthouse-core/audits/service-worker.js#L65-L67

so perhaps `ss` and `ssA/B` ?

**Pull request merged**